nightfall-optimist/src/services/database.mjs

```diff
@@ -419,35 +419,37 @@ export async function clearBlockNumberL1ForTransaction(transactionHashL1) {
 // and is part of an L2 block
 export async function getL2TransactionByCommitment(
   commitmentHash,
-  inL2AndNotInL2 = false,
+  { isAlreadyInL2, isAlreadyInMempool },
   blockNumberL2OfTx,
 ) {
   const connection = await mongo.connection(MONGO_URL);
   const db = connection.db(OPTIMIST_DB);
-  const query = inL2AndNotInL2
-    ? { commitments: commitmentHash }
-    : {
-        commitments: commitmentHash,
-        blockNumberL2: { $gte: -1, $ne: blockNumberL2OfTx },
-      };
+  const query =
+    isAlreadyInL2 && isAlreadyInMempool
+      ? { commitments: commitmentHash }
+      : {
+          commitments: commitmentHash,
+          blockNumberL2: { $gte: -1, $ne: blockNumberL2OfTx },
+        };
```

Comment:
I still find that isAlreadyInL2 and isAlreadyMempool names are confusing. What the function is doing is checking if the transaction can be found in L2 or in the mempool. Therefore, I would suggest changing the names for more clear ones such as "checkL2" and "checkMempool". 

On the other hand, I don't fully understand how the query is built. From my understanding, there's no scenario in which the function will be called with both flags set to false. However, by default both parameters are set to false. Moreover, right now it is always checking if it is in L2 and optionally if it is also in mempool. Therefore, the else condition is not only handling the case true / false but also two additional undesired cases.

I think it would probably be easier to remove "isAlreadyInL2" flag. If so, then probably a more appropiate name for the other flag would be "alsoCheckMempool". Then, if in a future PR we want to just check if a transaction is in the mempool and not in L2 we can create a different function that only checks for a duplicate tx in the mempool.

Reply:
hi @RogerTaule 
so flag "isAlreadyInL2" and "isAlreadyInMempool" are only used in two functions `getL2TransactionByCommitment` and `getL2TransactionByNullifier` and the purpose of both of these function is to query transaction collection by commitment and nullifier respectively.
And for now, functions `getL2TransactionByCommitment` and `getL2TransactionByNullifier` are only called in checkTransaction (functions checkDuplicateCommitment and checkDuplicateNullifier) specifically to fetch tx if there any which holds same commitment or nullifier

@ChaitanyaKonda  please correct if I am wrong
so understanding is user can submit multiple tx for same commitment (for better gas so that current proposer can pick and submit, but this not possible for now by checkTransaction call at time of transaction submit). so any given time same commitment or nullifier in multiple tx can present in mempool and one such tx in L2.

Note: `removeTransactionsFromMemPool` is called twice (in block assembly and in block proposed

checkTransaction are called in four case
1. Transaction submit
        isAlreadyInL2 = true & isAlreadyInMempool = true    (regardless where tx is fetch if it exist)
2. block assembly
        isAlreadyInL2 = true & isAlreadyInMempool = false  (look tx only in L2)
3. block proposed
        isAlreadyInL2 = true & isAlreadyInMempool = false  (look tx only in L2)
4. rollback
         isAlreadyInL2 = true & isAlreadyInMempool = false (look tx only in L2)

Reply:
So, I understand the logic, but since you are always checking in L2 and the only that changes is if you are checking also the mempool, why do we need the first flag? I mean, it would be enough with a flag that specifies if you want to check just in L2 or also check the mempool. I think it makes more sense done this way because the case false false will never exist and currently there's no case in which you are checking for a tx in the mempool but not in L2. If in a future this case arises, we could simply create another function.

I still think those names are very confusing. isAlreadyInL2 doesn't mean that the tx is in L2 but that you are looking IF the transaction is already in L2. Therefore, as suggested in my above comment, I would change the names with others such as checkL2 or checkMempool

Reply:
I kind of feel `check` prefix to variable name is usually should go for function name which does the check job.
And I have used `is` prefix for the boolean flags
@ChaitanyaKonda can we have your take on it.

regarding case [false  & false]  i also believe such case never will be handled.
but for [false & true] case when we want all/one tx in mempool of particular (commitment or nullifier)
apologies @ChaitanyaKonda  since you have created this issue and suggested these changes
can we have your thought on this as well whenever you are free